### PR TITLE
[flinksink] Add base implementation for writing files to the FS

### PR DESCRIPTION
**Pull request by pkubit-g**

### CODE REVIEW

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSink.java [hunk 1]
```diff
@@ -78,20 +79,31 @@
...
     private final DeltaSinkBuilder<IN> sinkBuilder;
 
     DeltaSink(DeltaSinkBuilder<IN> sinkBuilder) {
         this.sinkBuilder = checkNotNull(sinkBuilder);
     }
 
+    /**
+     * This method creates the {@link SinkWriter} instance that will be responsible for passing
+     * incoming stream events to the correct bucket writer and then flushed to the underlying files.
+     * <p>
```
Comment (scottsand-db) [inline flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSink.java]: nit: no need for this `<p>` here, unless in future PRs there will be more text below.

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSink.java [hunk 2]
```diff
@@ -78,20 +79,31 @@
...
+     *
+     * @param context {@link SinkWriter} init context object
+     * @param states  restored states of the writers. Will be empty collection for fresh start.
+     * @return new {@link SinkWriter} object
+     * @throws IOException Thrown, if the recoverable writer cannot be instantiated.
```
Comment (scottsand-db) [inline flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSink.java]: nit: don't think we need the extra word `Thrown`

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSink.java [hunk 3]
```diff
@@ -145,15 +157,50 @@
...
+        return new DeltaSinkBuilder.DefaultDeltaFormatBuilder<>(
+            basePath,
+            conf,
+            writerFactory
+        );
+    }
+
+    /**
+     * Convenience method for creating {@link DeltaSink}.
+     * Developer must ensure that writerFactory is configured to use SNAPPY compression codec.
```
Comment (scottsand-db) [inline flink-connector/src/main/java/org/apache/flink/connector/delta/sink/DeltaSink.java:187]: just so you know, when we generate javadocs, newlines are removed and only `<p>` actually provides a new line in the output. So
```
Convenience method for creating {@link DeltaSink}.
Developer must ensure that writerFactory is configured to use SNAPPY compression codec.
```

just becomes

```
Convenience method for creating {@link DeltaSink}. Developer must ensure that writerFactory is configured to use SNAPPY compression codec.
```

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriter.java [hunk 4]
```diff
@@ -45,32 +59,310 @@
...
 
-    public DeltaWriter() {
+    private static final Logger LOG = LoggerFactory.getLogger(DeltaWriter.class);
+
+    ///////////////////////////////////////////////////////////////////////////
+    // FileSink-specific fields
+    ///////////////////////////////////////////////////////////////////////////
+
+    ///////////////////////////////////////////////////////////////////////////
+    // configuration fields
```
Comment (scottsand-db) [inline flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriter.java:83]: if `configuration fields` is a "sub-header" of `FileSink-specific fields`, can you make the `\\\\\\\\...` shorter?
i.e.
```
///////////////////////////////////////////////////////////////////////////
// FileSink-specific fields
///////////////////////////////////////////////////////////////////////////
...
///////////////////////////////////////////////////
// configuration fields
///////////////////////////////////////////////////
```

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriterBucket.java [hunk 5]
```diff
@@ -57,39 +70,313 @@
...
+     * except that it uses custom implementation to represent the in-progress part file.
+     * See {@link DeltaInProgressPart} for details.
+     */
+    private DeltaInProgressPart<IN> rollPartFile(long currentTime) throws IOException {
+        closePartFile();
+
+        final Path partFilePath = assembleNewPartPath();
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(
```
Comment (scottsand-db) [inline flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriterBucket.java:232]: nit: is there no way to move these `LOG.isDebugEnabled` checks to some common `Logging` class?
i.e. most codes just say `logDebug(...)` and then the Logging class performs the `LOG.isDebugEnabled` check for us.

**Review by scottsand-db — COMMENTED** [review]
Did a first scan - will do more. Looks great! Great documentation - really helps with the readability.

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriter.java [hunk 6]
```diff
@@ -45,32 +59,310 @@
...
 
+    /**
+     * A proxy method that forwards the incoming event to the correct {@link DeltaWriterBucket}
+     * instance.
+     *
+     * @param element incoming stream event
+     * @param context context for getting additional data about input event
+     * @implNote This method behaves in the same way as
+     * {@link org.apache.flink.connector.file.sink.writer.FileWriter#write}
+     * except that it uses custom {@link DeltaWriterBucket} implementation.
```
Comment (tdas) [inline flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriter.java:190]: add why it uses the custom one

File: flink-connector/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/DeltaBulkBucketWriter.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+import java.io.IOException;
+
+import org.apache.flink.api.common.serialization.BulkWriter;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.core.fs.RecoverableFsDataOutputStream;
+import org.apache.flink.core.fs.RecoverableWriter;
+import org.apache.flink.util.Preconditions;
+
+/**
+ * This class is provided as a part of workaround for getting actual file size.
```
Comment (tdas) [inline flink-connector/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/DeltaBulkBucketWriter.java:32]: not clear what this class does. the reason it exists does not quite make sense without more information abouut what it does

Comment (pkubit-g) [reply flink-connector/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/DeltaBulkBucketWriter.java:32]: actually it is explained below:
```
Compared to its original version {@link BulkPartWriter} it changes only the return types
for methods {@link this#resumeFrom} and {@link this#openNew} to a custom implementation
of {@link BulkPartWriter} that is {@link DeltaBulkPartWriter}.
 ```
 
So we need this class only to modify the objects being returned from mentioned functions. And the actual workaround is described and implemented in `DeltaBulkPartWriter`.

Another thing is why we need it at all but to understand this one must get familiar with classes belonging to the package `org.apache.flink.streaming.api.functions.sink.filesystem` and `org.apache.flink.core.fs` in Flink (eg. `flink-file-sink-common` module which we are utilizing in out implementation). Those packages/modules/classes are like a generic ones for interacting with file systems in Flink.
Long story short all of the interfaces are build around the concept of classes like `BucketWriter`, `InProgressFileWriter`, `BulkBucketWriter`,  `RecoverableWriter` and so on which are user-facing abstractions. Then during runtime Flink resolves the correct implementations to use like `org.apache.flink.runtime.fs.hdfs.HadoopRecoverableWriter` or `org.apache.flink.core.fs.local.LocalRecoverableWriter` but on a sink-implementation level it's a black box for us.

File: flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriterBucket.java (hunk 5 above)
Comment (pkubit-g) [reply flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriterBucket.java:232]: I was just following Flink's style of coding here where they are not using any helper methods, but sure we can apply one.
But can you tell what's the concept behind and the full interface of proposed `logDebug` method  ? 
As far as i understand `isDebugEnabled` logic here it exists so as not to materialize the String's objects if they will not be used. But if we move the check and the actual logging to a helper method then we will be creating those objects on every method's call which will impact latency.

Comment (scottsand-db) [reply flink-connector/src/main/java/org/apache/flink/connector/delta/sink/writer/DeltaWriterBucket.java:232]: Great point! Then this as it is is fine :)
